File: cs/Scenarios/Client.NodeExplorer/Form1.cs
// Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

namespace Client.NodeExplorer
{
    using System;
    using System.Windows.Forms;

    using Opc.UaFx;
    using Opc.UaFx.Client;

    public partial class Form1 : Form
    {
        private readonly OpcClient client;


        public Form1()
            : base()
        {
            this.client = new OpcClient();
            this.InitializeComponent();
        }

        protected override void OnFormClosing(FormClosingEventArgs e)
        {
            this.client.Disconnect();
            base.OnFormClosing(e);
        }

        private bool Browse()
        {
            var result = false;

            try {
                var node = this.client.BrowseNode(OpcObjectTypes.RootFolder);
                result = this.Browse(node);
            }
            catch (OpcException ex) {
                this.ShowMessage("Browse", "Failed to browse: " + ex.Message);
            }

            return result;
        }

        private bool Browse(OpcNodeInfo node)
        {
            this.nodesTreeView.Nodes.Clear();
            return this.Browse(node, this.nodesTreeView.Nodes);
        }

        private bool Browse(OpcNodeInfo node, TreeNodeCollection treeNodes)
        {
            var result = false;

            try {
                var treeNode = treeNodes.Add(node.DisplayName.Value);

                if (node is OpcObjectNodeInfo) {
                    treeNode.ImageIndex = 0;

                    if (node.Reference.TypeDefinitionId == Opc.Ua.ObjectTypeIds.FolderType)
                        treeNode.ImageIndex = 3;
                    if (node.Reference.TypeDefinitionId == Opc.Ua.ObjectTypeIds.FileType)
                        treeNode.ImageIndex = 4;
                }
                else if (node is OpcMethodNodeInfo) {
                    treeNode.ImageIndex = 1;
                }
                else if (node is OpcVariableNodeInfo) {
                    treeNode.ImageIndex = 2;

                    if (node.Reference.TypeDefinitionId == Opc.Ua.VariableTypeIds.PropertyType)
                        treeNode.ImageIndex = 5;
                }

                treeNode.Tag = node;

                treeNode.Nodes.Add("Browsing...");
                result = true;
            }
            catch (OpcException ex) {
                this.ShowMessage("Browse", "Failed to browse: " + ex.Message);
            }

            return result;
        }

        private bool Connect()
        {
            var result = false;

            try {
                this.client.Connect();
                result = true;
            }
            catch (OpcException ex) {
                this.ShowMessage("Connect", "Failed to connect: " + ex.Message);
            }

            return result;
        }

        private void HandleConnectButtonClick(object sender, EventArgs e)
        {
            this.client.Disconnect();

            if (Uri.TryCreate(this.serverAddressTextBox.Text, UriKind.Absolute, out var serverAddress)) {
                this.client.ServerAddress = serverAddress;

                if (this.Connect())
                    this.Browse();
            }
            else {
                this.ShowMessage("Connect", "Invalid server address.");
            }
        }

        private void HandleNodesTreeViewAfterExpand(object sender, TreeViewEventArgs e)
        {
            if (e.Node.Tag is OpcNodeInfo node) {
                var treeNodes = e.Node.Nodes;
                treeNodes.Clear();

                foreach (var childNode in node.Children()) {
                    if (!this.Browse(childNode, treeNodes))
                        break;
                }
            }
        }

        private void ShowMessage(string caption, string text)
        {
            MessageBox.Show(
                    owner: this,
                    text,
                    caption,
                    MessageBoxButtons.OK,
                    MessageBoxIcon.Information);
        }
    }
}

File: vb/Scenarios/Client.NodeExplorer/Form1.vb
' Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

Imports Opc.UaFx
Imports Opc.UaFx.Client

Public Partial Class Form1
    Private ReadOnly client As OpcClient

    Public Sub New()
        MyBase.New()
        Me.client = New OpcClient()
        Me.InitializeComponent()
    End Sub

    Protected Overrides Sub OnFormClosing(ByVal e As FormClosingEventArgs)
        Me.client.Disconnect()
        MyBase.OnFormClosing(e)
    End Sub

    Private Function Browse() As Boolean
        Dim result = False

        Try
            Dim node = Me.client.BrowseNode(OpcObjectTypes.RootFolder)
            result = Me.Browse(node)
        Catch ex As OpcException
            Me.ShowMessage("Browse", "Failed to browse: " & ex.Message)
        End Try

        Return result
    End Function

    Private Function Browse(ByVal node As OpcNodeInfo) As Boolean
        Me.nodesTreeView.Nodes.Clear()
        Return Me.Browse(node, Me.nodesTreeView.Nodes)
    End Function

    Private Function Browse(
            ByVal node As OpcNodeInfo,
            ByVal treeNodes As TreeNodeCollection) As Boolean
        Dim result = False

        Try
            Dim treeNode = treeNodes.Add(node.DisplayName.Value)

            If TypeOf node Is OpcObjectNodeInfo Then
                treeNode.ImageIndex = 0
                If node.Reference.TypeDefinitionId = Opc.Ua.ObjectTypeIds.FolderType Then
                    treeNode.ImageIndex = 1
                End If
            ElseIf TypeOf node Is OpcMethodNodeInfo Then
                treeNode.ImageIndex = 2
            ElseIf TypeOf node Is OpcVariableNodeInfo Then
                treeNode.ImageIndex = 3
                If node.Reference.ReferenceType = OpcReferenceType.HasProperty Then
                    treeNode.ImageIndex = 4
                End If
            End If

            treeNode.Tag = node
            treeNode.Nodes.Add("Browsing...")

            result = True
        Catch ex As OpcException
            Me.ShowMessage("Browse", "Failed to browse: " & ex.Message)
        End Try

        Return result
    End Function

    Private Function Connect() As Boolean
        Dim result = False

        Try
            Me.client.Connect()
            result = True
        Catch ex As OpcException
            Me.ShowMessage("Connect", "Failed to connect: " & ex.Message)
        End Try

        Return result
    End Function

    Private Sub HandleConnectButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles connectButton.Click
        Me.client.Disconnect()
        Dim serverAddress As Uri = Nothing

        If Uri.TryCreate(Me.serverAddressTextBox.Text, UriKind.Absolute, serverAddress) Then
            Me.client.ServerAddress = serverAddress
            If Me.Connect() Then Me.Browse()
        Else
            Me.ShowMessage("Connect", "Invalid server address.")
        End If
    End Sub

    Private Sub HandleNodesTreeViewAfterExpand(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles nodesTreeView.AfterExpand
        Dim node As OpcNodeInfo = TryCast(e.Node.Tag, OpcNodeInfo)

        If node IsNot Nothing Then
            Dim treeNodes = e.Node.Nodes
            treeNodes.Clear()

            For Each childNode In node.Children()
                If Not Me.Browse(childNode, treeNodes) Then Exit For
            Next
        End If
    End Sub

    Private Sub ShowMessage(ByVal caption As String, ByVal text As String)
        MessageBox.Show(owner:=Me, text, caption, MessageBoxButtons.OK, MessageBoxIcon.Information)
    End Sub
End Class
